### PR TITLE
Add support for intrinsics (189)

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for 1st review from @arporter

File: src/fparser/two/Fortran2003.py [hunk 1]
```diff
@@ -9234,6 +9243,293 @@ def match(string):
...
+        result = CallBase.match(
+            Intrinsic_Name, Actual_Arg_Spec_List, string)
+        if result:
+            # There is a match so check the number of args provided
+            # matches the number of args expected by the intrinsic.
+            function_name = str(result[0])
+            function_args = result[1]
+            # This if/else will not be needed once issue #170 has been
+            # addressed.
+            if (isinstance(function_args, Actual_Arg_Spec_List)):
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py]: pylint complains about unnecessary parens here

File: src/fparser/two/Fortran2003.py [hunk 2]
```diff
@@ -9234,6 +9243,293 @@ def match(string):
...
+
+            min_nargs = Intrinsic_Name.generic_function_names[test_name]["min"]
+            max_nargs = Intrinsic_Name.generic_function_names[test_name]["max"]
+
+            if min_nargs == max_nargs and nargs != min_nargs:
+                raise InternalSyntaxError(
+                    "Intrinsic '{0}' expects {1} arg(s) but found {2}."
+                    "".format(function_name, min_nargs, nargs))
+            if min_nargs < max_nargs and (nargs < min_nargs or
+               nargs > max_nargs):
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py]: pylint complains about the indentation here.

**Comment by arporter** [comment]
I'm using this branch in https://github.com/stfc/psyclone/issue/339 and have realised I need to differentiate between unary, binary and nary intrinsics. Rather than construct separate lists in PSyclone, would it make sense to do that here? You already have the necessary information - it's just a matter of looping through and constructing the lists.

**Comment by arporter** [comment]
An alternative, which would make the code in PSyclone nicer, would be to have Unary, Binary and Nary sub-classes of `Intrinsic_Function_Statement` and create the appropriate one depending on how many arguments the intrinsic function is being passed.

**Comment by arporter** [comment]
I've discovered a bug. If I do:

    reader = FortranStringReader(
        "ze_z = SUM( e1t(:,:) * e2t(:,:) * zav_tide(:,:,jk) * "
        "tmask_i(:,:) ) &\n"
        "   &  / MAX( 1.e-20, SUM( e1t(:,:) * e2t(:,:) * wmask (:,:,jk) * "
        "tmask_i(:,:) ) )")

then the fparser2 node representing the `MAX` contains the following in `items`:

    (Intrinsic_Name('MAX'), Actual_Arg_Spec_List(',', (Real_Literal_Constant('1.E-20', None), Part_Ref(Name('SUM'), Add_Operand(Add_Operand(Add_Operand(Part_Ref(Name('e1t'), Section_Subscript_List(',', (Subscript_Triplet(None, None, None), Subscript_Triplet(None, None, None)))), '*', Part_Ref(Name('e2t'), Section_Subscript_List(',', (Subscript_Triplet(None, None, None), Subscript_Triplet(None, None, None))))), '*', Part_Ref(Name('wmask'), Section_Subscript_List(',', (Subscript_Triplet(None, None, None), Subscript_Triplet(None, None, None), Name('jk'))))), '*', Part_Ref(Name('tmask_i'), Section_Subscript_List(',', (Subscript_Triplet(None, None, None), Subscript_Triplet(None, None, None)))))))))

i.e. the `SUM` that is a child of `MAX` is represented by a `Part_Ref`, not an `Intrinsic_Function_Reference`.

File: doc/developers_guide.rst [hunk 3]
```diff
@@ -568,25 +568,24 @@ arguments. The first argument is a reader object which allows the line
...
-level `Program` class. However, this exception is not able to give any
-details of the error as is knows nothing about which rules failed to
+Currently the main use of `FortranSyntaxError` is to catch either an
+`InternalSyntaxError` exception or the final `NoMatchError` exception
+and re-raise it with line number and the text of the line to be
+output. These exceptions are caught and re-raised by overriding the
+`Base` class `__new__` method in the top level `Program` class. A
+limitation of the `NoMatchError` exception (but not the
+`InternalSyntaxError` exception) is that it is not able to give any
+details of the error, as is knows nothing about which rules failed to
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/developers_guide.rst]: oops

File: doc/developers_guide.rst [hunk 4]
```diff
@@ -568,25 +568,24 @@ arguments. The first argument is a reader object which allows the line
...
-a match, but that the match is known to have a syntax error. This
-approach leads to good quality feedback to the user on the type of
-error and its location and should be used wherever possible. One issue
-is that when `FortranSyntaxError` is raised from one of these
-additional places the fparser script is not able to use the reader's
-fifo buffer to extract position information. This is dealt with by not
-outputting anything from the script related to the fifo buffer in this
-case. It is possible that if the lines were pushed back into the
-buffer then this would work.
+`FortranSyntaxError` should also be used when it is know that there is
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/developers_guide.rst]: oops again.

File: doc/developers_guide.rst [hunk 5]
```diff
@@ -568,25 +568,24 @@ arguments. The first argument is a reader object which allows the line
...
+a match, the match has a syntax error and the line number information
+is available via the reader object. One issue is that when
+`FortranSyntaxError` is raised from such a location, the fparser
+script is not able to use the reader's fifo buffer to extract position
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/developers_guide.rst]: OK, I've tried to clarify - given my partial knowledge of what the problem is.

File: src/fparser/two/Fortran2003.py [hunk 6]
```diff
@@ -215,6 +215,8 @@ def __new__(cls, string):
             # At the moment there is no useful information provided by
             # NoMatchError so we pass on an empty string.
             raise FortranSyntaxError(string, "")
+        except InternalSyntaxError as excinfo:
+            raise FortranSyntaxError(string, excinfo)
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: Done.

File: src/fparser/two/Fortran2003.py [hunk 7]
```diff
@@ -4323,8 +4325,15 @@ class Primary(Base):  # R701
             or type-param-name
             or ( expr )
 
+    `intrinsic_function_reference` is not part of rule R701 but is
+    required for fparser to recognise intrinsic functions. This should
+    be placed before array-constructor in the list so that an
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: Yes, thanks.

File: src/fparser/two/Fortran2003.py [hunk 8]
```diff
@@ -9234,6 +9243,293 @@ def match(string):
...
+        "MVBITS": {"min": 5, "max": 5},
+        "NOT": {"min": 1, "max": 1}}
+
+    floating_point_manipulation_names = {
+        "EXPONENT": {"min": 1, "max": 1},
+        "FRACTION": {"min": 1, "max": 1},
+        "NEAREST": {"min": 2, "max": 2},
+        "RRSPACING": {"min": 1, "max": 1},
+        "SCALE": {"min": 2, "max": 2},
+        "SET EXPONENT": {"min": 2, "max": 2},
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: Well spotted.

File: src/fparser/two/Fortran2003.py [hunk 9]
```diff
@@ -9234,6 +9243,293 @@ def match(string):
...
+        "CCOS": "COS", "CEXP": "EXP", "CLOG": "LOG", "CSIN": "SIN",
+        "CSQRT": "SQRT", "DABS": "ABS", "DACOS": "ACOS", "DASIN": "ASIN",
+        "DATAN": "ATAN", "DATAN2": "ATAN2", "DCOS": "COS", "DCOSH": "COSH",
+        "DDIM": "DIM", "DEXP": "EXP", "DINT": "AINT", "DLOG": "LOG",
+        "DLOG10": "LOG10", "DMAX1": "MAX", "DMIN1": "MIN", "DMOD": "MOD",
+        "DNINT": "ANINT", "DSIGN": "SIGN", "DSIN": "SIN", "DSINH": "SINH",
+        "DSQRT": "SQRT", "DTAN": "TAN", "DTANH": "TANH", "FLOAT": "REAL",
+        "IABS": "ABS", "IDIM": "DIM", "IDINT": "INT", "IDNINT": "NINT",
+        "IFIX": "INT", "ISIGN": "SIGN", "MAX0": "MAX", "MAX1": "MAX",
+        "MIN0": "MIN", "MIN1": "MIN1", "SNGL": "REAL"}
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: That's a great spot!

File: src/fparser/two/Fortran2003.py [hunk 10]
```diff
@@ -9234,6 +9243,293 @@ def match(string):
...
+        '''Attempt to match the input `string` with the intrinsic function
+        names defined in `generic_function_names` or
+        `specific_function_names`. If there is a match the resultant
+        string will be converted to upper case.
+
+        :param str string: The pattern to be matched.
+
+        :returns: A tuple containing the matched string (converted to \
+        upper case) if there is a match or None if there is not.
+        :rtype: (str) or NoneType
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: Oops.

File: src/fparser/two/Fortran2003.py [hunk 11]
```diff
@@ -9234,6 +9243,293 @@ def match(string):
...
+                raise InternalSyntaxError(
+                    "Intrinsic '{0}' expects {1} arg(s) but found {2}."
+                    "".format(function_name, min_nargs, nargs))
+            if min_nargs < max_nargs and (nargs < min_nargs or
+               nargs > max_nargs):
+                raise InternalSyntaxError(
+                    "Intrinsic '{0}' expects between {1} and {2} args but "
+                    "found {3}.".format(function_name, min_nargs, max_nargs,
+                                        nargs))
+            if max_nargs == -1 and nargs < min_nargs:
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: I agree. Changed.

File: src/fparser/two/tests/fortran2003/test_intrinsics.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
+# Copyright (c) 2019 Science and Technology Facilities Council
+
+# All rights reserved.
+
+# Modifications made as part of the fparser project are distributed
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_intrinsics.py]: Good point. Other new files are also incorrect, but that is for another issue.

File: src/fparser/two/tests/fortran2003/test_intrinsics.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_intrinsics.py]: There is no authorship in fparser at the moment.

File: src/fparser/two/tests/fortran2003/test_intrinsics.py [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+    intrinsic syntax error.
+
+    '''
+    reader = get_reader("subroutine sub()\na = sin(b,c)\nend subroutine sub\n")
+    with pytest.raises(FortranSyntaxError) as excinfo:
+        _ = Program(reader)
+    assert (
+        "at line 2\n"
+        ">>>a = sin(b,c)\n"
+        "Intrinsic 'SIN' expects 1 arg(s) but found 2") in str(excinfo.value)
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_intrinsics.py]: :-)

File: src/fparser/two/tests/fortran2003/test_intrinsics.py [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+
+def test_intrinsic_name_specific(f2003_create):
+    '''Test that class Intrinsic_Name correctly matches a specific name.'''
+    result = Intrinsic_Name("CCOS")
+    assert isinstance(result, Intrinsic_Name)
+    assert str(result) == "CCOS"
+
+
+def test_intrinsic_name_invalid(f2003_create):
+    '''Test that class Intrinsic_Name correctly matches a specific name.'''
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_intrinsics.py]: Well spotted again.

File: src/fparser/two/tests/fortran2003/test_primary_r701.py [hunk 16]
```diff
@@ -109,6 +109,15 @@ def possible_subclasses(node_type, _seen=None):
     return seen
 
 
+def test_intrinsic_function(f2003_create):
+    '''Test that an intrinsic functions are matched by Primary.
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_primary_r701.py]: Eagle eyed!

File: src/fparser/two/Fortran2003.py (hunk 1 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: Modified.

File: src/fparser/two/Fortran2003.py (hunk 2 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: Fixed. Sorry, I'm not sure how the pylint errors got through.

**Comment by rupertford** [comment]
> I'm using this branch in [stfc/PSyclone#339](https://github.com/stfc/PSyclone/issues/339) and have realised I need to differentiate between unary, binary and nary intrinsics. Rather than construct separate lists in PSyclone, would it make sense to do that here? You already have the necessary information - it's just a matter of looping through and constructing the lists.

My gut feeling is that this is not an issue for fparser. It doesn't care about unary, binary and nary so these should be constructed outside.

**Comment by rupertford** [comment]
> I've discovered a bug. If I do:
> 
> ```
> reader = FortranStringReader(
>     "ze_z = SUM( e1t(:,:) * e2t(:,:) * zav_tide(:,:,jk) * "
>     "tmask_i(:,:) ) &\n"
>     "   &  / MAX( 1.e-20, SUM( e1t(:,:) * e2t(:,:) * wmask (:,:,jk) * "
>     "tmask_i(:,:) ) )")
> ```
> 
> then the fparser2 node representing the `MAX` contains the following in `items`:
> 
> ```
> (Intrinsic_Name('MAX'), Actual_Arg_Spec_List(',', (Real_Literal_Constant('1.E-20', None), Part_Ref(Name('SUM'), Add_Operand(Add_Operand(Add_Operand(Part_Ref(Name('e1t'), Section_Subscript_List(',', (Subscript_Triplet(None, None, None), Subscript_Triplet(None, None, None)))), '*', Part_Ref(Name('e2t'), Section_Subscript_List(',', (Subscript_Triplet(None, None, None), Subscript_Triplet(None, None, None))))), '*', Part_Ref(Name('wmask'), Section_Subscript_List(',', (Subscript_Triplet(None, None, None), Subscript_Triplet(None, None, None), Name('jk'))))), '*', Part_Ref(Name('tmask_i'), Section_Subscript_List(',', (Subscript_Triplet(None, None, None), Subscript_Triplet(None, None, None)))))))))
> ```
> 
> i.e. the `SUM` that is a child of `MAX` is represented by a `Part_Ref`, not an `Intrinsic_Function_Reference`.

I can recreate the observed problem (intrinsic in the argument list of an intrinsic) so have added a failing test which I will fix in this issue. I will also tests to make sure that intrinsics in other situations also get recognised.

**Comment by rupertford** [comment]
The bug is now fixed and I've hopefully addressed the reviewers comments. Ready for next review from @arporter

File: src/fparser/two/Fortran2003.py [hunk 17]
```diff
@@ -9234,6 +9248,301 @@ def match(string):
...
+            min_nargs = Intrinsic_Name.generic_function_names[test_name]["min"]
+            max_nargs = Intrinsic_Name.generic_function_names[test_name]["max"]
+
+            if max_nargs is None:
+                if nargs < min_nargs:
+                    # None indicates an unlimited number of arguments
+                    raise InternalSyntaxError(
+                        "Intrinsic '{0}' expects at least {1} args but found "
+                        "{2}.".format(function_name, min_nargs, nargs))
+                else:
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py]: pylint says this `else` is unnecessary.

File: src/fparser/scripts/fparser2.py [hunk 18]
```diff
@@ -97,13 +97,15 @@ def runner(_, options, args):
         try:
             f2003_parser = ParserFactory().create()
             program = f2003_parser(reader)
+            # print (repr(program))
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/fparser/scripts/fparser2.py]: Oops, this should have been removed. With it removed the fparser2.py script should be unchanged so, as just discussed, I will not update the docstring. This is not laziness as PR #200 adds support for outputting the `repr()` version so should update docstrings appropriately.

File: src/fparser/two/Fortran2003.py (hunk 17 above)
Comment (rupertford) [reply src/fparser/two/Fortran2003.py]: And pylint is correct :-).

**Comment by rupertford** [comment]
Ready for next review from @arporter. I've got a good feeling ...

**Pull request merged**